traffic_ops/traffic_ops_golang/iso/iso.go

```diff
@@ -0,0 +1,368 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-rfc"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/jmoiron/sqlx"
+)
+
+// Various directories and filenames related to ISO generation.
+const (
+	cfgDefaultDir   = "/var/www/files"  // Default directory containing config file
+	cfgFilename     = "osversions.json" // The JSON config file containing mapping of OS names to directories
+	cfgFilenamePerl = "osversions.cfg"  // Config file name in the Perl version
+
+	// This is the directory name inside each OS directory where
+	// configuration files for kickstart scripts are placed.
+	ksCfgDir = "ks_scripts"
+
+	// Configuration files that are generated inside the ks_scripts directory.
+	ksCfgNetwork     = "network.cfg"
+	ksCfgMgmtNetwork = "mgmt_network.cfg"
+	ksCfgPassword    = "password.cfg"
+	ksCfgDisk        = "disk.cfg"
+	ksStateOut       = "state.out"
+
+	ksAltCommand = "generate" // Optional executable that is invoked instead of mkisofs
+)
+
+// Various database columns and values.
+const (
+	ksFilesParamName       = "kickstart.files.location"
+	ksFilesParamConfigFile = "mkisofs"
+)
+
+// Various HTTP-related values.
+const (
+	httpHeaderContentDisposition = "Content-Disposition"
+	httpHeaderContentType        = "Content-Type"
+	httpHeaderContentDownload    = "application/download"
+)
+
+// ISOs handler is responsible for generating and returning an ISO image,
+// as a streaming download.
+//
+// Response types:
+//
+// Error:
+//   HTTP 400
+//   {
+//     "alerts": [
+//       {"level":"error","text":"hostName is required"},
+//       {"level":"error","text":"disk is required"},
+//       ...,
+//     ]
+//   }
+//
+// Success:
+//   HTTP 200
+//   Content-Disposition: attachment; filename="db.infra.ciab.test-centos72.iso"
+//   Content-Type: application/download
+//
+func ISOs(w http.ResponseWriter, req *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(req, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, req, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	// Decode request body into isoRequest instance.
+	var ir isoRequest
+	if err := json.NewDecoder(req.Body).Decode(&ir); err != nil {
```

Comment:
Since `isoRequest` already has a validation method, why not make it return an `error` instead of a `string` and accept a `*database/sql.Tx` (which would be unused, I know) and then it could be an [`api.ParseValidator`](https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api#ParseValidator) and then you could use [`api.Parse`](https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api#Parse)